botocore/utils.py

```diff
@@ -621,6 +622,26 @@ def fix_s3_host(request, signature_version, region_name, **kwargs):
     addressing.  This allows us to avoid 301 redirects for all
     bucket names that can be CNAME'd.
     """
+    # By default we do not use virtual hosted style addressing when
+    # signed with signature version 4.
+    if signature_version in ['s3v4', 'v4']:
+        return
+    try:
+        switch_to_virtual_host_style(
+            request, signature_version, region_name, 's3.amazonaws.com')
+    except InvalidDNSNameError as e:
+        bucket_name = e.kwargs['bucket_name']
+        logger.debug('Not changing URI, bucket is not DNS compatible: %s',
+                     bucket_name)
+
+
+def switch_to_virtual_host_style(request, signature_version, region_name,
```

Comment:
Can you document the params?  Not sure what default_endpoint_url is suppose to be.